File: python/src/acedatacloud/resources/maestro.py
"""Maestro AI Video Studio resources (``/maestro/*``)."""

from __future__ import annotations

from typing import Any

from acedatacloud._runtime.tasks import AsyncTaskHandle, TaskHandle


class Maestro:
    """Synchronous Maestro client."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport

    def videos(
        self,
        *,
        prompt: str,
        action: str | None = None,
        ref_task_id: str | None = None,
        file_urls: list[str] | None = None,
        langs: list[str] | None = None,
        aspect: str | None = None,
        duration: int | None = None,
        quality: str | None = None,
        scenario: str | None = None,
        style: str | None = None,
        voice: str | None = None,
        callback_url: str | None = None,
        wait: bool = False,
        poll_interval: float = 5.0,
        max_wait: float = 600.0,
        **kwargs: Any,
    ) -> dict[str, Any] | TaskHandle:
        body: dict[str, Any] = {"prompt": prompt, **kwargs}
        if action is not None:
            body["action"] = action
        if ref_task_id is not None:
            body["ref_task_id"] = ref_task_id
        if file_urls is not None:
            body["file_urls"] = file_urls
        if langs is not None:
            body["langs"] = langs
        if aspect is not None:
            body["aspect"] = aspect
        if duration is not None:
            body["duration"] = duration
        if quality is not None:
            body["quality"] = quality
        if scenario is not None:
            body["scenario"] = scenario
        if style is not None:
            body["style"] = style
        if voice is not None:
            body["voice"] = voice
        if callback_url is not None:
            body["callback_url"] = callback_url

        result = self._transport.request("POST", "/maestro/videos", json=body)
        task_id = result.get("task_id")

        if not task_id or not wait:
            return result

        handle = TaskHandle(task_id, "/maestro/tasks", self._transport)
        if wait:
            return handle.wait(poll_interval=poll_interval, max_wait=max_wait)
        return handle


class AsyncMaestro:
    """Async Maestro client."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport

    async def videos(
        self,
        *,
        prompt: str,
        action: str | None = None,
        ref_task_id: str | None = None,
        file_urls: list[str] | None = None,
        langs: list[str] | None = None,
        aspect: str | None = None,
        duration: int | None = None,
        quality: str | None = None,
        scenario: str | None = None,
        style: str | None = None,
        voice: str | None = None,
        callback_url: str | None = None,
        wait: bool = False,
        poll_interval: float = 5.0,
        max_wait: float = 600.0,
        **kwargs: Any,
    ) -> dict[str, Any] | AsyncTaskHandle:
        body: dict[str, Any] = {"prompt": prompt, **kwargs}
        if action is not None:
            body["action"] = action
        if ref_task_id is not None:
            body["ref_task_id"] = ref_task_id
        if file_urls is not None:
            body["file_urls"] = file_urls
        if langs is not None:
            body["langs"] = langs
        if aspect is not None:
            body["aspect"] = aspect
        if duration is not None:
            body["duration"] = duration
        if quality is not None:
            body["quality"] = quality
        if scenario is not None:
            body["scenario"] = scenario
        if style is not None:
            body["style"] = style
        if voice is not None:
            body["voice"] = voice
        if callback_url is not None:
            body["callback_url"] = callback_url

        result = await self._transport.request("POST", "/maestro/videos", json=body)
        task_id = result.get("task_id")

        if not task_id or not wait:
            return result

        handle = AsyncTaskHandle(task_id, "/maestro/tasks", self._transport)
        if wait:
            return await handle.wait(poll_interval=poll_interval, max_wait=max_wait)
        return handle

File: typescript/src/resources/maestro.ts
/** Maestro AI Video Studio resources (`/maestro/*`). */

import { Transport } from '../runtime/transport';
import { TaskHandle } from '../runtime/tasks';

export class Maestro {
  constructor(private transport: Transport) {}

  async videos(opts: {
    prompt: string;
    action?: 'generate' | 'remix' | 'edit' | 'extend' | (string & {});
    refTaskId?: string;
    fileUrls?: string[];
    langs?: string[];
    aspect?: '9:16' | '16:9' | '1:1' | (string & {});
    duration?: number;
    quality?: 'draft' | 'standard' | 'premium' | (string & {});
    scenario?: 'auto' | 'narrated' | 'drama' | 'avatar' | 'motion' | 'slideshow' | (string & {});
    style?: string;
    voice?: string;
    callbackUrl?: string;
    wait?: boolean;
    pollInterval?: number;
    maxWait?: number;
    [key: string]: unknown;
  }): Promise<Record<string, unknown> | TaskHandle> {
    const { prompt, action, refTaskId, fileUrls, langs, aspect, duration, quality, scenario, style, voice, callbackUrl, wait: shouldWait, pollInterval, maxWait, ...rest } = opts;
    const body: Record<string, unknown> = { prompt, ...rest };
    if (action !== undefined) body.action = action;
    if (refTaskId !== undefined) body.ref_task_id = refTaskId;
    if (fileUrls !== undefined) body.file_urls = fileUrls;
    if (langs !== undefined) body.langs = langs;
    if (aspect !== undefined) body.aspect = aspect;
    if (duration !== undefined) body.duration = duration;
    if (quality !== undefined) body.quality = quality;
    if (scenario !== undefined) body.scenario = scenario;
    if (style !== undefined) body.style = style;
    if (voice !== undefined) body.voice = voice;
    if (callbackUrl !== undefined) body.callback_url = callbackUrl;

    const result = await this.transport.request('POST', '/maestro/videos', { json: body });
    const taskId = result.task_id as string | undefined;

    if (!taskId || !shouldWait) return result;

    const handle = new TaskHandle(taskId, '/maestro/tasks', this.transport);
    if (shouldWait) return handle.wait({ pollInterval, maxWait });
    return handle;
  }
}
